src/sage/graphs/matching_covered_graph.py

```diff
@@ -0,0 +1,2070 @@
+            raise ValueError('an odd no. of distinct vertices can not be '
+                             'removed from a matching covered graph')
+
+        for vertex in vertices:
+            if vertex not in self:
+                raise ValueError('vertex (%s) not in the graph' % str(vertex))
+
+        if self.order() == len(vertices):
+            raise ValueError('the resulting graph after the removal of the '
+                             'vertices is trivial, therefore is not '
+                             'matching covered')
+
+        try:
+            G = Graph(self, multiedges=self.allows_multiple_edges())
+            G.delete_vertices(vertices)
+
+            M = Graph(self.get_matching())
+
+            M.delete_vertices(vertices)
+            # The resulting matching after the removal of the input vertices
+            # must be a valid perfect matching of the resulting graph obtained
+            # after the removal of the vertices
+
+            if (G.order() != 2*M.size()):
+                M = None
+
+            self.__init__(data=G, matching=M)
+
+        except Exception:
+            raise ValueError('the resulting graph after the removal of '
+                             'the vertices is not matching covered')
+
+    @doc_index('Overwritten methods')
+    def get_matching(self):
+        r"""
+        Return a :class:`~EdgesView` of ``self._matching`` (a perfect matching
+        of the (matching covered) graph computed at the initialization).
+
+        OUTPUT:
+
+        - This method returns :class:`EdgesView` of the edges of a
+          perfect matching of the (matching covered) graph.
+
+        EXAMPLES:
+
+        If one specifies a perfect matching while initializing the object, the
+        value of ``self._matching`` is captures the same matching::
+
+            sage: P = graphs.PetersenGraph()
+            sage: M = [(0, 1), (2, 3), (4, 9), (5, 7), (6, 8)]
+            sage: G = MatchingCoveredGraph(P, M)
+            sage: sorted(G.get_matching())
+            [(0, 1), (2, 3), (4, 9), (5, 7), (6, 8)]
+            sage: M == sorted(G.get_matching())
+            True
+
+        If no matching is specified while initilizing a matching covered graph,
+        a perfect is computed
+        :meth:`~sage.graphs.graph.Graph.matching` and that is captured as
+        ``self._matching``::
+
+            sage: P = graphs.PetersenGraph()
+            sage: M = P.matching()
+            sage: G = MatchingCoveredGraph(P)
+            sage: sorted(G.get_matching())
+            [(0, 5, None), (1, 6, None), (2, 7, None), (3, 8, None), (4, 9, None)]
+            sage: sorted(M) == sorted(G.get_matching())
+            True
+        """
+        return self._matching
+
+    @doc_index('Barriers and canonical partition')
+    def maximal_barrier(self, vertex):
+        r"""
+        Return the (unique) maximal barrier containing the vertex.
+
+        We use the following theorem.
```

Comment:
Before explaining the theorem you use, add a like explaining what is a `barrier`